### PR TITLE
Refresh tokens metadata

### PR DESCRIPTION
**Pull request by letier3110**
https://www.notion.so/madfissolutions/Refresh-metadata-7b17e2a058f141d9b01a17fbcd5c8892

### CODE REVIEW

**Review by LendiHop — APPROVED** [review]

File: src/store/wallet/wallet-epics.ts [hunk 1]
```diff
@@ -114,6 +119,18 @@ const loadTokenMetadataEpic = (action$: Observable<Action>) =>
     )
   );
 
+const loadTokensMetadataEpic = (action$: Observable<Action>, state$: Observable<RootState>) =>
+  action$.pipe(
+    ofType(loadTokensMetadataActions.submit),
+    withTokenList(state$),
+    switchMap(tokens =>
+      loadTokensMetadata$(Object.keys(tokens)).pipe(
```
Comment (unixvb) [inline src/store/wallet/wallet-epics.ts:127]: load metadata only for "Unknown Token"s

File: src/store/wallet/wallet-actions.ts [hunk 2]
```diff
@@ -30,6 +30,10 @@ export const loadTokenMetadataActions = createActions<
   string
 >('assets/LOAD_TOKEN_METADATA');
 
+export const loadTokensMetadataActions = createActions<void, Array<TokenMetadataInterface>, string>(
```
Comment (unixvb) [inline src/store/wallet/wallet-actions.ts:33]: rename to `loadUnknownTokensMetadataActions` (and align other names with it)

File: src/store/wallet/wallet-epics.ts [hunk 3]
```diff
@@ -114,6 +119,22 @@ const loadTokenMetadataEpic = (action$: Observable<Action>) =>
...
 
+const loadTokensMetadataEpic = (action$: Observable<Action>, state$: Observable<RootState>) =>
+  action$.pipe(
+    ofType(loadTokensMetadataActions.submit),
+    withTokenList(state$),
+    switchMap(tokens =>
+      loadTokensMetadata$(
+        Object.values(tokens)
+          .filter(x => x.symbol === '???')
+          .map(getTokenSlug)
```
Comment (unixvb) [inline src/store/wallet/wallet-epics.ts:130]: create separate util `getUnknownTokensSlugs`

File: src/utils/token-metadata.utils.ts [hunk 4]
```diff
@@ -32,6 +33,43 @@ export const loadTokenMetadata$ = memoize(
...
 
+type DetailedAssetMetdata = {
+  decimals: number;
+  symbol: string;
+  name: string;
+  shouldPreferSymbol?: boolean;
+  thumbnailUri?: string;
+  displayUri?: string;
+  artifactUri?: string;
+};
```
Comment (unixvb) [inline src/utils/token-metadata.utils.ts:44]: Take a look at `TokenMetadataResponse`.  
`tokenMetadataApi.get<TokenMetadataResponse>(/metadata/${address}/${id} `
returns the same type that 
`tokenMetadataApi.post<Array<DetailedAssetMetdata | null>>('/', slugs) does`

Fix `TokenMetadataResponse` interface and reuse it

File: src/utils/token-metadata.utils.ts [hunk 5]
```diff
@@ -32,6 +33,43 @@ export const loadTokenMetadata$ = memoize(
...
+
+          return {
+            id: Number(slugs[index].split('_')[1]),
+            address: slugs[index].split('_')[0],
+            decimals: token.decimals,
+            symbol: token.symbol ?? token.name?.substring(0, 8) ?? '???',
+            name: token.name ?? token.symbol ?? 'Unknown Token',
+            thumbnailUri: token.thumbnailUri,
+            artifactUri: token.artifactUri
+          };
```
Comment (unixvb) [inline src/utils/token-metadata.utils.ts:63]: create shared with `loadTokenMetadata$` util that will create correct metadata

File: src/utils/token-metadata.utils.ts [hunk 6]
```diff
@@ -16,22 +17,45 @@ export interface TokenMetadataResponse {
...
-      }))
+      map(({ data }) => transformDataToTokenMetadata(data, address, id)),
+      filter(isDefined)
     ),
   { cacheKey: ([address, id]) => getTokenSlug({ address, id }) }
 );
 
+export const loadTokensMetadata$ = memoize(
+  (slugs: Array<string>): Observable<Array<TokenMetadataInterface>> =>
+    from(tokenMetadataApi.post<Array<TokenMetadataInterface | null>>('/', slugs)).pipe(
```
Comment (unixvb) [inline src/utils/token-metadata.utils.ts:48]: mb TokenMetadataResponse should be here.  
if so, change transformDataToTokenMetadata

File: src/store/wallet/wallet-actions.ts [hunk 7]
```diff
@@ -30,6 +30,10 @@ export const loadTokenMetadataActions = createActions<
   string
 >('assets/LOAD_TOKEN_METADATA');
 
+export const loadUnknownTokensMetadataActions = createActions<void, Array<TokenMetadataInterface>, string>(
+  'assets/LOAD_TOKENS_METADATA'
```
Comment (unixvb) [inline src/store/wallet/wallet-actions.ts:34]: change action type

**Review by unixvb — APPROVED** [review]

**Pull request merged**